### PR TITLE
Matrix Cleanup Phase 1

### PR DESCRIPTION
**Pull request by siefkenj**
This pull request is the first phase in getting https://github.com/sympy/sympy/pull/11592 accepted.

PR Contents
----------------
This PR splits many of the functions from `BaseMatrix` into separate classes based on category.  They include

- `MatrixShaping` which handles all reshaping, row joins, etc. operations
- `MatrixProperties` which handles all the basic `is_` properties and methods but leaves the advanced ones like `is_diagonalizable` out.
- `MatrixOperations` which includes `transpose`, `subs`, `applyfunc` and friends that do some sort of element-wise operation to the matrix and return a new matrix result.

Each of these inherit from `MatrixRequired` which has stubs indicating each method required of a new matrix.  The methods implemented in the new Matrix classes have been removed from `MatrixBase` and `MatrixBase` now inherits from each of them.

Tests for each of the new Matrix classes have been added in `test_commonmatrix.py`.  These tests were mostly copied from `test_matrices.py`.  Additionally, a `_MinimalMatrix` class has been added that implements everything `MatrixRequired` needs (plus `__repr__` and a more-advanced-than-needed `__init__` function).  `_MinimalMatrix` is used in the unit tests to ensure that each of the new Matrix classes works on its own and doesn't need any additional functionality.

Each new Matrix class has public methods with argument checking/sanitizing and then defers to private `_eval_...` methods to do the computation.  Subclasses can override these `_eval_` methods to get specialized behavior (for example for `SparseMatrix`).  This means the behavior can be specialized while keeping the docstrings and bounds-checking, etc. around from the base classes.

Future
---------
It seems like there are two or three more natural categories to break up the methods in `BaseMatrix`.  These include a class that handles all matrix arithmetic, a class that handles determinants (and the various ways to compute them), a class that handles row reduction and solving matrix equations, and a class that handles eigen computations.  These classes get progressively higher level and will inherit from `MatrixOperations`, `MatrixProperties`, and `MatrixShaping` as required (rather than only inheriting from `MatrixRequired`).

Once these are implemented, the remainder of https://github.com/sympy/sympy/pull/11592 should be a lot easier to review, as then it will be mostly clean-up of the dense and sparse matrix modules.

### CODE REVIEW

**Comment by moorepants** [comment]
Please merge in master to resolve the conflicts.

**Comment by siefkenj** [comment]
@moorepants done

**Comment by moorepants** [comment]
Can you report the coverage of these tests for your new code?

**Comment by siefkenj** [comment]
@moorepants The tests in `test_commonmatrix.py` cover every method that doesn't start with`_` in the new classes I added.  Is this what you're asking?

**Comment by moorepants** [comment]
I'm asking specifically for the results of `bin/coverage_report.py` for these tests.

**Comment by siefkenj** [comment]
@moorepants I was unware of that file or what it does, so I am not sure how to answer your question...

**Comment by moorepants** [comment]
You can run that file like you run the tests and it will produce a coverage report showing how well the tests exercise each line of code in the respective modules.

**Comment by siefkenj** [comment]
@moorepants  I ran it on `test_commonmatrix.py` and got

    Coverage for sympy/matrices/tests/test_commonmatrix.py : 99%

Is that what you wanted?  I'm also not sure about this python 3.5 error that showed up.  I don't have python 3.5 on my system :-/

File: sympy/matrices/matrices.py [hunk 1]
```diff
@@ -73,2747 +76,3386 @@ def __repr__(self):
...
         ========
 
-        berkowitz
+        >>> from sympy.abc import x
+        >>> from sympy.matrices import Matrix
+        >>> Matrix([[x], [1]]).free_symbols
+        set([x])
         """
-        return roots(self.berkowitz_charpoly(Dummy('x')), **flags)
+        return self._eval_free_symbols()
```
Comment (moorepants) [inline sympy/matrices/matrices.py]: What is the reason to have this code in a separate `_eval_*` method? Seems like creating two methods when only one is needed.

Comment (moorepants) [inline sympy/matrices/matrices.py]: Ok, I see your explanation:

> Each new Matrix class has public methods with argument checking/sanitizing and then defers to private _eval_... methods to do the computation. Subclasses can override these _eval_ methods to get specialized behavior (for example for SparseMatrix). This means the behavior can be specialized while keeping the docstrings and bounds-checking, etc. around from the base classes.

This seems fine.

**Comment by moorepants** [comment]
I'm +1 to merge this. I think it does help organize the vary large Matrix code. But it is at the expense of using more advanced features of Python that are not as easily understood by beginners. I think that's an ok tradeoff.

**Pull request merged**